File: c#/src/base/Exchange.Types.cs
namespace Main;

//
// Exchange Types that are used in the Exchange.Wrappers.cs file to type the generic methods
//

public partial class Exchange
{

    public struct Precision
    {
        public float amount;
        public float price;
        public Precision(object precision2)
        {
            var precision = (Dictionary<string, object>)precision2;
            amount = precision.ContainsKey("amount") ? (float)precision["amount"] : 0;
            price = precision.ContainsKey("price") ? (float)precision["price"] : 0;
        }
    }

    public struct MinMax
    {
        public float min;
        public float max;
        public MinMax(object minMax2)
        {
            var minMax = (Dictionary<string, object>)minMax2;
            min = minMax.ContainsKey("min") ? (float)minMax["min"] : 0;
            max = minMax.ContainsKey("max") ? (float)minMax["max"] : 0;
        }
    }

    public struct Fee
    {
        public float rate;
        public float cost;

        public Fee(object fee2)
        {
            var fee = (Dictionary<string, object>)fee2;
            rate = fee.ContainsKey("rate") ? (float)fee["rate"] : 0;
            cost = fee.ContainsKey("cost") ? (float)fee["cost"] : 0;
        }
    }

    public struct Limits
    {
        public MinMax? amount;
        public MinMax? cost;
        public MinMax? leverage;
        public MinMax? price;

        public Limits(object limits2)
        {
            var limits = (Dictionary<string, object>)limits2;
            amount = limits.ContainsKey("amount") ? new MinMax(limits["amount"]) : null;
            cost = limits.ContainsKey("cost") ? new MinMax(limits["cost"]) : null;
            leverage = limits.ContainsKey("leverage") ? new MinMax(limits["leverage"]) : null;
            price = limits.ContainsKey("price") ? new MinMax(limits["price"]) : null;
        }
    }

    public struct Market
    {
        public string? id;
        public string? symbol;
        public string? baseCurrency;
        public string? quote;
        public string? baseId;
        public string? quoteId;
        public bool? active;
        public string? type;
        public bool? spot;
        public bool? margin;
        public bool? swap;
        public bool? future;
        public bool? option;
        public bool? contract;
        public string? settle;
        public string? settleId;
        public float contractSize;
        public bool? linear;
        public bool? inverse;
        public float expiry;
        public string? expiryDatetime;
        public float strike;
        public string? optionType;
        public float taker;
        public float maker;
        public bool? percentage;
        public bool? tierBased;
        public string? feeSide;

        public Precision? precision;

        public Limits? limits;
        public Dictionary<string, object> info;

        public Market(object market2)
        {
            var market = (Dictionary<string, object>)market2;
            id = market.ContainsKey("id") ? (string)market["id"] : null;
            symbol = market.ContainsKey("symbol") ? (string)market["symbol"] : null;
            baseCurrency = market.ContainsKey("base") ? (string)market["base"] : null;
            quote = market.ContainsKey("quote") ? (string)market["quote"] : null;
            baseId = market.ContainsKey("baseId") ? (string)market["baseId"] : null;
            quoteId = market.ContainsKey("quoteId") ? (string)market["quoteId"] : null;
            active = market.ContainsKey("active") ? (bool)market["active"] : null;
            type = market.ContainsKey("type") ? (string)market["type"] : null;
            spot = market.ContainsKey("spot") ? (bool)market["spot"] : null;
            margin = market.ContainsKey("margin") ? (bool)market["margin"] : null;
            swap = market.ContainsKey("swap") ? (bool)market["swap"] : null;
            future = market.ContainsKey("future") ? (bool)market["future"] : null;
            option = market.ContainsKey("option") ? (bool)market["option"] : null;
            contract = market.ContainsKey("contract") ? (bool)market["contract"] : null;
            settle = market.ContainsKey("settle") ? (string)market["settle"] : null;
            settleId = market.ContainsKey("settleId") ? (string)market["settleId"] : null;
            contractSize = market.ContainsKey("contractSize") ? (float)market["contractSize"] : 0;
            linear = market.ContainsKey("linear") ? (bool)market["linear"] : null;
            inverse = market.ContainsKey("inverse") ? (bool)market["inverse"] : null;
            expiry = market.ContainsKey("expiry") ? (float)market["expiry"] : 0;
            expiryDatetime = market.ContainsKey("expiryDatetime") ? (string)market["expiryDatetime"] : null;
            strike = market.ContainsKey("strike") ? (float)market["strike"] : 0;
            optionType = market.ContainsKey("optionType") ? (string)market["optionType"] : null;
            taker = market.ContainsKey("taker") ? (float)market["taker"] : 0;
            maker = market.ContainsKey("maker") ? (float)market["maker"] : 0;
            percentage = market.ContainsKey("percentage") ? (bool)market["percentage"] : null;
            tierBased = market.ContainsKey("tierBased") ? (bool)market["tierBased"] : null;
            feeSide = market.ContainsKey("feeSide") ? (string)market["feeSide"] : null;
            precision = market.ContainsKey("precision") ? new Precision(market["precision"]) : null;
            limits = market.ContainsKey("limits") ? new Limits(market["limits"]) : null;
            info = market;
        }
    }

    public struct Trade
    {
        public float amount;
        public float price;
        public float cost;
        public string? id;
        public string? orderId;
        public Dictionary<string, object>? info;
        public Int64? timestamp;
        public string? datetime;
        public string? symbol;
        public string? type;
        public string? side;
        public string? takerOrMaker;
        public Fee? fee;
        public Trade(object trade2)
        {
            var trade = (Dictionary<string, object>)trade2;
            amount = trade.ContainsKey("amount") ? (float)trade["amount"] : 0;
            price = trade.ContainsKey("price") ? (float)trade["price"] : 0;
            cost = trade.ContainsKey("cost") ? (float)trade["cost"] : 0;
            id = trade.ContainsKey("id") ? (string)trade["id"] : null;
            orderId = trade.ContainsKey("orderId") ? (string)trade["orderId"] : null;
            info = trade.ContainsKey("info") ? (Dictionary<string, object>)trade["info"] : null;
            timestamp = trade.ContainsKey("timestamp") ? (Int64)trade["timestamp"] : null;
            datetime = trade.ContainsKey("datetime") ? (string)trade["datetime"] : null;
            symbol = trade.ContainsKey("symbol") ? (string)trade["symbol"] : null;
            type = trade.ContainsKey("type") ? (string)trade["type"] : null;
            side = trade.ContainsKey("side") ? (string)trade["side"] : null;
            takerOrMaker = trade.ContainsKey("takerOrMaker") ? (string)trade["takerOrMaker"] : null;
            fee = trade.ContainsKey("fee") ? new Fee(trade["fee"]) : null;
        }
    }

    public struct Order
    {
        public string? id;
        public string? clientOrderId;
        public Int64? timestamp;
        public string? datetime;
        public string? lastTradeTimestamp;
        public string? symbol;
        public string? type;
        public string? side;
        public string? price;
        public float? cost;
        public float? average;
        public float? amount;
        public float? filled;
        public float? remaining;
        public string? status;
        public Fee? fee;
        public IEnumerable<Trade>? trades;
        public Dictionary<string, object>? info;
        public Order(object order2)
        {
            var order = (Dictionary<string, object>)order2;
            id = order.ContainsKey("id") ? (string)order["id"] : null;
            clientOrderId = order.ContainsKey("clientOrderId") ? (string)order["clientOrderId"] : null;
            timestamp = order.ContainsKey("timestamp") ? (Int64)order["timestamp"] : null;
            datetime = order.ContainsKey("datetime") ? (string)order["datetime"] : null;
            lastTradeTimestamp = order.ContainsKey("lastTradeTimestamp") ? (string)order["lastTradeTimestamp"] : null;
            symbol = order.ContainsKey("symbol") ? (string)order["symbol"] : null;
            type = order.ContainsKey("type") ? (string)order["type"] : null;
            side = order.ContainsKey("side") ? (string)order["side"] : null;
            price = order.ContainsKey("price") ? (string)order["price"] : null;
            cost = order.ContainsKey("cost") ? (float)order["cost"] : null;
            average = order.ContainsKey("average") ? (float)order["average"] : null;
            amount = order.ContainsKey("amount") ? (float)order["amount"] : null;
            filled = order.ContainsKey("filled") ? (float)order["filled"] : null;
            remaining = order.ContainsKey("remaining") ? (float)order["remaining"] : null;
            status = order.ContainsKey("status") ? (string)order["status"] : null;
            fee = order.ContainsKey("fee") ? new Fee(order["fee"]) : null;
            trades = order.ContainsKey("trades") ? ((IEnumerable<object>)order["trades"]).Select(x => new Trade(x)) : null;
            info = order.ContainsKey("info") ? (Dictionary<string, object>)order["info"] : null;
        }
    }

    public struct Ticker
    {
        public string? symbol;
        public Int64 timestamp;
        public string? datetime;
        public float high;
        public float low;
        public float bid;
        public float bidVolume;
        public float ask;
        public float askVolume;
        public float vwap;
        public float open;

        public float close;
        public float last;
        public float previousClose;
        public float change;
        public float percentage;
        public float average;
        public float baseVolume;
        public float quoteVolume;

        public Ticker(object ticker2)
        {
            var ticker = (Dictionary<string, object>)ticker2;
            symbol = ticker.ContainsKey("symbol") ? (string)ticker["symbol"] : null;
            timestamp = ticker.ContainsKey("timestamp") ? (Int64)ticker["timestamp"] : 0;
            datetime = ticker.ContainsKey("datetime") ? (string)ticker["datetime"] : null;
            high = ticker.ContainsKey("high") ? (float)ticker["high"] : 0;
            low = ticker.ContainsKey("low") ? (float)ticker["low"] : 0;
            bid = ticker.ContainsKey("bid") ? (float)ticker["bid"] : 0;
            bidVolume = ticker.ContainsKey("bidVolume") ? (float)ticker["bidVolume"] : 0;
            ask = ticker.ContainsKey("ask") ? (float)ticker["ask"] : 0;
            askVolume = ticker.ContainsKey("askVolume") ? (float)ticker["askVolume"] : 0;
            vwap = ticker.ContainsKey("vwap") ? (float)ticker["vwap"] : 0;
            open = ticker.ContainsKey("open") ? (float)ticker["open"] : 0;
            close = ticker.ContainsKey("close") ? (float)ticker["close"] : 0;
            last = ticker.ContainsKey("last") ? (float)ticker["last"] : 0;
            previousClose = ticker.ContainsKey("previousClose") ? (float)ticker["previousClose"] : 0;
            change = ticker.ContainsKey("change") ? (float)ticker["change"] : 0;
            percentage = ticker.ContainsKey("percentage") ? (float)ticker["percentage"] : 0;
            average = ticker.ContainsKey("average") ? (float)ticker["average"] : 0;
            baseVolume = ticker.ContainsKey("baseVolume") ? (float)ticker["baseVolume"] : 0;
            quoteVolume = ticker.ContainsKey("quoteVolume") ? (float)ticker["quoteVolume"] : 0;
        }

        struct Transaction
        {
            public string? id;
            public string? txid;
            public string? address;
            public string? tag;
            public string? type;
            public string? currency;
            public float amount;
            public string? status;
            public Int64? updated;
            public Int64? timestamp;
            public string? datetime;

            public Transaction(object transaction2)
            {
                var transaction = (Dictionary<string, object>)transaction2;
                id = transaction.ContainsKey("id") ? (string)transaction["id"] : null;
                txid = transaction.ContainsKey("txid") ? (string)transaction["txid"] : null;
                address = transaction.ContainsKey("address") ? (string)transaction["address"] : null;
                tag = transaction.ContainsKey("tag") ? (string)transaction["tag"] : null;
                type = transaction.ContainsKey("type") ? (string)transaction["type"] : null;
                currency = transaction.ContainsKey("currency") ? (string)transaction["currency"] : null;
                amount = transaction.ContainsKey("amount") ? (float)transaction["amount"] : 0;
                status = transaction.ContainsKey("status") ? (string)transaction["status"] : null;
                updated = transaction.ContainsKey("updated") ? (Int64)transaction["updated"] : null;
                timestamp = transaction.ContainsKey("timestamp") ? (Int64)transaction["timestamp"] : null;
                datetime = transaction.ContainsKey("datetime") ? (string)transaction["datetime"] : null;
            }
        }
    }

    public struct OrderBook
    {
        public List<List<float>>? bids;
        public List<List<float>>? asks;

        public string? symbol;
        public Int64 timestamp;
        public string? datetime;
        public Int64? nonce;

        public OrderBook(object orderbook2)
        {
            var orderbook = (Dictionary<string, object>)orderbook2;
            bids = orderbook.ContainsKey("bids") ? ((IEnumerable<object>)orderbook["bids"]).Select(x => ((IEnumerable<object>)x).Select(y => (float)y).ToList()).ToList() : null;
            asks = orderbook.ContainsKey("asks") ? ((IEnumerable<object>)orderbook["asks"]).Select(x => ((IEnumerable<object>)x).Select(y => (float)y).ToList()).ToList() : null;
            symbol = orderbook.ContainsKey("symbol") ? (string)orderbook["symbol"] : null;
            timestamp = orderbook.ContainsKey("timestamp") ? (Int64)orderbook["timestamp"] : 0;
            datetime = orderbook.ContainsKey("datetime") ? (string)orderbook["datetime"] : null;
            nonce = orderbook.ContainsKey("nonce") ? (Int64)orderbook["nonce"] : null;
        }
    }

    public struct OHLCV
    {
        public Int64 timestamp;
        public float open;
        public float high;
        public float low;
        public float close;
        public float volume;

        public OHLCV(object ohlcv2)
        {
            var ohlcv = (List<object>)ohlcv2;
            timestamp = (Int64)ohlcv[0];
            open = (float)ohlcv[1];
            high = (float)ohlcv[2];
            low = (float)ohlcv[3];
            close = (float)ohlcv[4];
            volume = (float)ohlcv[5];
        }
    }

    public struct OHLCVC
    {
        public Int64 timestamp;
        public float open;
        public float high;
        public float low;
        public float close;
        public float volume;
        public float cost;

        public OHLCVC(object ohlcv2)
        {
            var ohlcv = (List<object>)ohlcv2;
            timestamp = (Int64)ohlcv[0];
            open = (float)ohlcv[1];
            high = (float)ohlcv[2];
            low = (float)ohlcv[3];
            close = (float)ohlcv[4];
            volume = (float)ohlcv[5];
            cost = (float)ohlcv[6];
        }
    }
}

File: go/ccxt/exchange_types.go
package ccxt

import (
	"fmt"
	"math"
	"time"
)

// Utility functions for safe extraction from maps
func SafeFloatTyped(m interface{}, key interface{}) float64 {
	res := SafeFloat(m, key, nil)
	if res != nil {
		return res.(float64)
	}
	return math.NaN()
}

func SafeStringTyped(m interface{}, key interface{}) string {
	res := SafeString(m, key, nil)
	if res != nil {
		return res.(string)
	}
	return ""
}

func SafeBoolTyped(m interface{}, key interface{}) bool {
	res := SafeBool(m, key, false)
	if res != nil {
		return res.(bool)
	}
	return false
}

func SafeInt64Typed(m interface{}, key interface{}) int64 {
	res := SafeInteger(m, key, nil)
	if res != nil {
		return res.(int64)
	}
	return 0
}

// MarketInterface struct
type MarketInterface struct {
	Info           map[string]interface{}
	UppercaseID    string
	LowercaseID    string
	Symbol         string
	BaseCurrency   string
	QuoteCurrency  string
	BaseID         string
	QuoteID        string
	Active         bool
	Type           string
	Spot           bool
	Margin         bool
	Swap           bool
	Future         bool
	Option         bool
	Contract       bool
	Settle         string
	SettleID       string
	ContractSize   float64
	Linear         bool
	Inverse        bool
	Quanto         bool
	Expiry         int64
	ExpiryDatetime string
	Strike         float64
	OptionType     string
	Taker          float64
	Maker          float64
	Limits         Limits
	Created        int64
}

// CreateMarketInterface initializes the MarketInterface struct
func CreateMarketInterface(data interface{}) MarketInterface {
	m := data.(map[string]interface{})

	// Handle limits if present
	var limits Limits
	if v, ok := m["limits"]; ok {
		limits = CreateLimits(v)
	}

	return MarketInterface{
		Info:           m,
		UppercaseID:    SafeStringTyped(m, "uppercaseId"),
		LowercaseID:    SafeStringTyped(m, "lowercaseId"),
		Symbol:         SafeStringTyped(m, "symbol"),
		BaseCurrency:   SafeStringTyped(m, "base"),
		QuoteCurrency:  SafeStringTyped(m, "quote"),
		BaseID:         SafeStringTyped(m, "baseId"),
		QuoteID:        SafeStringTyped(m, "quoteId"),
		Active:         SafeBoolTyped(m, "active"),
		Type:           SafeStringTyped(m, "type"),
		Spot:           SafeBoolTyped(m, "spot"),
		Margin:         SafeBoolTyped(m, "margin"),
		Swap:           SafeBoolTyped(m, "swap"),
		Future:         SafeBoolTyped(m, "future"),
		Option:         SafeBoolTyped(m, "option"),
		Contract:       SafeBoolTyped(m, "contract"),
		Settle:         SafeStringTyped(m, "settle"),
		SettleID:       SafeStringTyped(m, "settleId"),
		ContractSize:   SafeFloatTyped(m, "contractSize"),
		Linear:         SafeBoolTyped(m, "linear"),
		Inverse:        SafeBoolTyped(m, "inverse"),
		Quanto:         SafeBoolTyped(m, "quanto"),
		Expiry:         SafeInt64Typed(m, "expiry"),
		ExpiryDatetime: SafeStringTyped(m, "expiryDatetime"),
		Strike:         SafeFloatTyped(m, "strike"),
		OptionType:     SafeStringTyped(m, "optionType"),
		Taker:          SafeFloatTyped(m, "taker"),
		Maker:          SafeFloatTyped(m, "maker"),
		Limits:         limits,
		Created:        SafeInt64Typed(m, "created"),
	}
}

// Precision struct
type Precision struct {
	Amount float64
	Price  float64
}

func CreatePrecision(data interface{}) Precision {
	m := data.(map[string]interface{})
	return Precision{
		Amount: SafeFloatTyped(m, "amount"),
		Price:  SafeFloatTyped(m, "price"),
	}
}

// MarketMarginModes struct
type MarketMarginModes struct {
	Cross    bool
	Isolated bool
}

func CreateMarketMarginModes(data interface{}) MarketMarginModes {
	m := data.(map[string]interface{})
	return MarketMarginModes{
		Cross:    SafeBoolTyped(m, "cross"),
		Isolated: SafeBoolTyped(m, "isolated"),
	}
}

// MinMax struct
type MinMax struct {
	Min float64
	Max float64
}

func CreateMinMax(data interface{}) MinMax {
	m := data.(map[string]interface{})
	return MinMax{
		Min: SafeFloatTyped(m, "min"),
		Max: SafeFloatTyped(m, "max"),
	}
}

// Fee struct
type Fee struct {
	Rate float64
	Cost float64
}

func CreateFee(data interface{}) Fee {
	m := data.(map[string]interface{})
	return Fee{
		Rate: SafeFloatTyped(m, "rate"),
		Cost: SafeFloatTyped(m, "cost"),
	}
}

// TradingFeeInterface struct
type TradingFeeInterface struct {
	Symbol     string
	Maker      float64
	Taker      float64
	Percentage bool
	TierBased  bool
	Info       map[string]interface{}
}

func CreateTradingFeeInterface(data interface{}) TradingFeeInterface {
	m := data.(map[string]interface{})
	return TradingFeeInterface{
		Symbol:     SafeStringTyped(m, "symbol"),
		Maker:      SafeFloatTyped(m, "maker"),
		Taker:      SafeFloatTyped(m, "taker"),
		Percentage: SafeBoolTyped(m, "percentage"),
		TierBased:  SafeBoolTyped(m, "tierBased"),
		Info:       m,
	}
}

// Limits struct
type Limits struct {
	Amount   *MinMax
	Cost     *MinMax
	Leverage *MinMax
	Price    *MinMax
}

func CreateLimits(data interface{}) Limits {
	m := data.(map[string]interface{})
	var amount, cost, leverage, price *MinMax
	if v, ok := m["amount"]; ok {
		amountValue := CreateMinMax(v)
		amount = &amountValue
	}
	if v, ok := m["cost"]; ok {
		costValue := CreateMinMax(v)
		cost = &costValue
	}
	if v, ok := m["leverage"]; ok {
		leverageValue := CreateMinMax(v)
		leverage = &leverageValue
	}
	if v, ok := m["price"]; ok {
		priceValue := CreateMinMax(v)
		price = &priceValue
	}
	return Limits{
		Amount:   amount,
		Cost:     cost,
		Leverage: leverage,
		Price:    price,
	}
}

// Market struct
type Market struct {
	ID            string
	Symbol        string
	BaseCurrency  string
	QuoteCurrency string
	BaseID        string
	QuoteID       string
	Active        bool
	Type          string
	Precision     *Precision
	MarginModes   *MarketMarginModes
	Limits        *Limits
	Info          map[string]interface{}
	Created       time.Time
}

func CreateMarket(data interface{}) Market {
	m := data.(map[string]interface{})
	var precision *Precision
	var marginModes *MarketMarginModes
	var limits *Limits
	if v, ok := m["precision"]; ok {
		precisionValue := CreatePrecision(v)
		precision = &precisionValue
	}
	if v, ok := m["marginModes"]; ok {
		marginModesValue := CreateMarketMarginModes(v)
		marginModes = &marginModesValue
	}
	if v, ok := m["limits"]; ok {
		limitsValue := CreateLimits(v)
		limits = &limitsValue
	}

	created := time.Unix(0, 0)
	if v, ok := m["created"]; ok {
		if timestamp, ok := v.(int64); ok {
			created = time.Unix(timestamp/1000, 0)
		}
	}

	return Market{
		ID:            SafeStringTyped(m, "id"),
		Symbol:        SafeStringTyped(m, "symbol"),
		BaseCurrency:  SafeStringTyped(m, "base"),
		QuoteCurrency: SafeStringTyped(m, "quote"),
		BaseID:        SafeStringTyped(m, "baseId"),
		QuoteID:       SafeStringTyped(m, "quoteId"),
		Active:        SafeBoolTyped(m, "active"),
		Type:          SafeStringTyped(m, "type"),
		Precision:     precision,
		MarginModes:   marginModes,
		Limits:        limits,
		Info:          m,
		Created:       created,
	}
}

// Trade struct
type Trade struct {
	Amount       float64
	Price        float64
	Cost         float64
	ID           string
	Order        string
	Info         map[string]interface{}
	Timestamp    int64
	Datetime     string
	Symbol       string
	Type         string
	Side         string
	TakerOrMaker string
	Fee          *Fee
}

func CreateTrade(data interface{}) Trade {
	m := data.(map[string]interface{})
	var fee *Fee
	if v, ok := m["fee"]; ok {
		feeValue := CreateFee(v)
		fee = &feeValue
	}

	return Trade{
		Amount:       SafeFloatTyped(m, "amount"),
		Price:        SafeFloatTyped(m, "price"),
		Cost:         SafeFloatTyped(m, "cost"),
		ID:           SafeStringTyped(m, "id"),
		Order:        SafeStringTyped(m, "order"),
		Timestamp:    SafeInt64Typed(m, "timestamp"),
		Datetime:     SafeStringTyped(m, "datetime"),
		Symbol:       SafeStringTyped(m, "symbol"),
		Type:         SafeStringTyped(m, "type"),
		Side:         SafeStringTyped(m, "side"),
		TakerOrMaker: SafeStringTyped(m, "takerOrMaker"),
		Fee:          fee,
		Info:         m,
	}
}

// Order struct
type Order struct {
	ID                 string
	ClientOrderID      string
	Timestamp          int64
	Datetime           string
	LastTradeTimestamp string
	Symbol             string
	Type               string
	Side               string
	Price              float64
	Cost               float64
	Average            float64
	Amount             float64
	Filled             float64
	Remaining          float64
	Status             string
	ReduceOnly         bool
	PostOnly           bool
	Fee                *Fee
	Trades             []Trade
	TriggerPrice       float64
	StopLossPrice      float64
	TakeProfitPrice    float64
	Info               map[string]interface{}
}

func CreateOrder(data interface{}) Order {
	m := data.(map[string]interface{})
	var fee *Fee
	if v, ok := m["fee"]; ok {
		feeValue := CreateFee(v)
		fee = &feeValue
	}
	var trades []Trade
	if v, ok := m["trades"]; ok {
		tradesData := v.([]interface{})
		for _, tradeData := range tradesData {
			trade := CreateTrade(tradeData)
			trades = append(trades, trade)
		}
	}

	return Order{
		ID:                 SafeStringTyped(m, "id"),
		ClientOrderID:      SafeStringTyped(m, "clientOrderId"),
		Timestamp:          SafeInt64Typed(m, "timestamp"),
		Datetime:           SafeStringTyped(m, "datetime"),
		LastTradeTimestamp: SafeStringTyped(m, "lastTradeTimestamp"),
		Symbol:             SafeStringTyped(m, "symbol"),
		Type:               SafeStringTyped(m, "type"),
		Side:               SafeStringTyped(m, "side"),
		Price:              SafeFloatTyped(m, "price"),
		Cost:               SafeFloatTyped(m, "cost"),
		Average:            SafeFloatTyped(m, "average"),
		Amount:             SafeFloatTyped(m, "amount"),
		Filled:             SafeFloatTyped(m, "filled"),
		Remaining:          SafeFloatTyped(m, "remaining"),
		Status:             SafeStringTyped(m, "status"),
		ReduceOnly:         SafeBoolTyped(m, "reduceOnly"),
		PostOnly:           SafeBoolTyped(m, "postOnly"),
		Fee:                fee,
		Trades:             trades,
		TriggerPrice:       SafeFloatTyped(m, "triggerPrice"),
		StopLossPrice:      SafeFloatTyped(m, "stopLossPrice"),
		TakeProfitPrice:    SafeFloatTyped(m, "takeProfitPrice"),
		Info:               m,
	}
}

// Ticker struct
type Ticker struct {
	Symbol        string
	Timestamp     int64
	Datetime      string
	High          float64
	Low           float64
	Bid           float64
	BidVolume     float64
	Ask           float64
	AskVolume     float64
	Vwap          float64
	Open          float64
	Close         float64
	Last          float64
	PreviousClose float64
	Change        float64
	Percentage    float64
	Average       float64
	BaseVolume    float64
	QuoteVolume   float64
	Info          map[string]interface{}
}

func CreateTicker(data interface{}) Ticker {
	m := data.(map[string]interface{})
	return Ticker{
		Symbol:        SafeStringTyped(m, "symbol"),
		Timestamp:     SafeInt64Typed(m, "timestamp"),
		Datetime:      SafeStringTyped(m, "datetime"),
		High:          SafeFloatTyped(m, "high"),
		Low:           SafeFloatTyped(m, "low"),
		Bid:           SafeFloatTyped(m, "bid"),
		BidVolume:     SafeFloatTyped(m, "bidVolume"),
		Ask:           SafeFloatTyped(m, "ask"),
		AskVolume:     SafeFloatTyped(m, "askVolume"),
		Vwap:          SafeFloatTyped(m, "vwap"),
		Open:          SafeFloatTyped(m, "open"),
		Close:         SafeFloatTyped(m, "close"),
		Last:          SafeFloatTyped(m, "last"),
		PreviousClose: SafeFloatTyped(m, "previousClose"),
		Change:        SafeFloatTyped(m, "change"),
		Percentage:    SafeFloatTyped(m, "percentage"),
		Average:       SafeFloatTyped(m, "average"),
		BaseVolume:    SafeFloatTyped(m, "baseVolume"),
		QuoteVolume:   SafeFloatTyped(m, "quoteVolume"),
		Info:          m,
	}
}

// OHLCV struct
type OHLCV struct {
	Timestamp int64
	Open      float64
	High      float64
	Low       float64
	Close     float64
	Volume    float64
}

func CreateOHLCV(data interface{}) OHLCV {
	ohlcv := data.([]interface{})
	return OHLCV{
		Timestamp: SafeInt64Typed(ohlcv, 0),
		Open:      SafeFloatTyped(ohlcv, 1),
		High:      SafeFloatTyped(ohlcv, 2),
		Low:       SafeFloatTyped(ohlcv, 3),
		Close:     SafeFloatTyped(ohlcv, 4),
		Volume:    SafeFloatTyped(ohlcv, 5),
	}
}

// transaction

type Transaction struct {
	ID        string
	TxID      string
	Address   string
	Tag       string
	Type      string
	Currency  string
	Amount    float64
	Status    string
	Updated   int64
	Timestamp int64
	Datetime  string
}

// NewTransaction initializes a Transaction struct from a map.
func CreateTransaction(transaction map[string]interface{}) Transaction {
	return Transaction{
		ID:        SafeStringTyped(transaction, "id"),
		TxID:      SafeStringTyped(transaction, "txid"),
		Address:   SafeStringTyped(transaction, "address"),
		Tag:       SafeStringTyped(transaction, "tag"),
		Type:      SafeStringTyped(transaction, "type"),
		Currency:  SafeStringTyped(transaction, "currency"),
		Amount:    SafeFloatTyped(transaction, "amount"),
		Status:    SafeStringTyped(transaction, "status"),
		Updated:   SafeInt64Typed(transaction, "updated"),
		Timestamp: SafeInt64Typed(transaction, "timestamp"),
		Datetime:  SafeStringTyped(transaction, "datetime"),
	}
}

// orderbook

type OrderBook struct {
	Bids      [][]float64
	Asks      [][]float64
	Symbol    string
	Timestamp int64
	Datetime  string
	Nonce     int64
}

// NewOrderBook initializes an OrderBook struct from a map.
func CreateOrderBook(orderbook2 interface{}) OrderBook {
	orderbook := orderbook2.(map[string]interface{})
	return OrderBook{
		Bids:      parseOrderBookEntries(orderbook, "bids"),
		Asks:      parseOrderBookEntries(orderbook, "asks"),
		Symbol:    SafeStringTyped(orderbook, "symbol"),
		Timestamp: SafeInt64Typed(orderbook, "timestamp"),
		Datetime:  SafeStringTyped(orderbook, "datetime"),
		Nonce:     SafeInt64Typed(orderbook, "nonce"),
	}
}

// parseOrderBookEntries extracts and converts order book entries to [][]float64.
func parseOrderBookEntries(orderbook map[string]interface{}, key string) [][]float64 {
	if value, ok := orderbook[key]; ok {
		if entries, ok := value.([]interface{}); ok {
			var result [][]float64
			for _, entry := range entries {
				if pair, ok := entry.([]interface{}); ok {
					var floatPair []float64
					for _, v := range pair {
						if num, ok := v.(float64); ok {
							floatPair = append(floatPair, num)
						}
					}
					if len(floatPair) == 2 { // Ensure bid/ask pairs have two values
						result = append(result, floatPair)
					}
				}
			}
			return result
		}
	}
	return nil
}

// tickers
// Tickers struct
type Tickers struct {
	Info    map[string]interface{}
	Tickers map[string]Ticker
}

// NewTickers initializes a Tickers struct from a map.
func CreateTickers(tickersData2 interface{}) Tickers {
	info := GetInfo(tickersData) // Assuming Helper.GetInfo is implemented as GetInfo
	tickersMap := make(map[string]Ticker)

	for key, value := range tickersData {
		if key != "info" {
			if tickerData, ok := value.(map[string]interface{}); ok {
				tickersMap[key] = CreateTicker(tickerData) // Assuming NewTicker is the Ticker constructor
			}
		}
	}

	return Tickers{
		Info:    info,
		Tickers: tickersMap,
	}
}

// GetTicker retrieves a Ticker by key, similar to the indexer in C#.
func (t *Tickers) GetTicker(key string) (Ticker, error) {
	ticker, exists := t.Tickers[key]
	if !exists {
		return Ticker{}, fmt.Errorf("the key '%s' was not found in the tickers", key)
	}
	return ticker, nil
}

// SetTicker sets or updates a Ticker by key.
func (t *Tickers) SetTicker(key string, ticker Ticker) {
	t.Tickers[key] = ticker
}

// Mocked GetInfo function for demonstration purposes.
func GetInfo(data map[string]interface{}) map[string]interface{} {
	if info, ok := data["info"].(map[string]interface{}); ok {
		return info
	}
	return nil
}
